### PR TITLE
docker: use HostConfig object when creating container with api_version > 1.15

### LINKED ISSUE #1158: Docker 1.6 cannot mount a volume with a single file
Issue Type: bug report
Ansible Version: 1.9.0.1
Environment: Ubuntu 14.04

**Summary:**
Docker allows mounting volumes specifying a single file, like `docker run -v /etc/profile:/etc/profile:ro busybox`
Starting from Docker version 1.6, Ansible cannot mount such volumes, if the file already exist within an image, but needs to be overwritten by a copy from the host.

This is probably a bug in Docker engine or`docker-py` library, but I can reproduce it only from Ansible.
Docker-compose has a similar issue: https://github.com/docker/compose/issues/424

**Steps To Reproduce:**
Run this playbook:

```

---
- hosts: 127.0.0.1
  connection: local
  tasks:
    - docker:
        image: busybox
        volumes:
        - "/etc/profile:/etc/profile:ro"
```

**Expected Results:**
Having Docker 1.5 it applies normally.

**Actual Results:**
With Docker 1.6 Ansible fails:

```
TASK: [docker ] ***************************************************************
failed: [127.0.0.1] => {"changed": false, "failed": true}
msg: Docker API Error: file exists at /var/lib/docker/aufs/mnt/f2ed605b182792d69c397b6dcc7d7c6c40193e89d49d047b33934f779423c38f/etc/profile, can't create volume there
```

### PR DESCRIPTION
**Pull request by ghost**
Fixes #1158 

Trying to use HostConfig if possible. This is recommended way as described in https://github.com/docker/docker-py/issues/546#issuecomment-96032262

### CODE REVIEW

**Comment by shin-** [comment]
Just to provide some context on that, the Docker Engine team has been pushing for everyone to provide `HostConfig` at create time rather than start time for a while now (the CLI does it exclusively, which is why docker-py users and ansible users by transition encounter those odd cases first). I'm strongly considering deprecating it in docker-py as well as I expect support for this feature to only get worse over time.

Related docker/docker issues:
- https://github.com/docker/docker/issues/10695
- https://github.com/docker/docker/issues/12759

**TL;DR: 100% support for this PR, in accord with the trend set by the Docker Engine project.**

**Comment by gregdek** [comment]
@mlosev Thanks for submitting this pull request. Please accept our apologies for taking so long to review it.

Unfortunately, due to this delay, it appears that the underlying code has changed, and this PR needs to be rebased. If you believe that this PR still applies, please rebase and we will then put it into our (new, faster) community review process as soon as possible.  Again, apologies for the delay and the extra work.

**Comment by tomjrob** [comment]
From my brief search in the issues, this pull request would fix a couple of the bugs i'm currently experiencing with docker support in Ansible; #1158, #1337. Additionally, I think this would provide the basis for fixing #1707 which I currently work around by pinning docker-py to v1.2.3.

Anyone (@mlosev?) have time to rebase this so it can go into the community review process as per @gregdek's comment?

If no one is already working on this, I'm happy to have a go over the weekend, but have not delved into the ansible code base before (new to the project) so I may need some guidance.

**Comment by larsks** [comment]
Other folks are getting hit by this as well.  [This question](http://stackoverflow.com/questions/31459513/ansible-and-docker-locally-build-image-get-pulled-and-causes-failure/31461925#31461925) cropped up on StackOverflow today.

**Comment by larsks** [comment]
I have rebased this against the current `devel` branch.

**Comment by bcoca** [comment]
yet it still has a conflict ....

**Comment by larsks** [comment]
I don't see it.  I can rebase on `devel` with no errors:

```
$ git rebase devel
Current branch bug/hostconfig is up to date.
```

And I can merge this into `origin/devel` with no errors:

```
$ git co -b integration origin/devel
Switched to a new branch 'integration'
$ git merge bug/hostconfig 
Updating fa85d95..efb6088
Fast-forward
 cloud/docker/docker.py | 62 +++++++++++++++++++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 55 insertions(+), 7 deletions(-)
```

I'm not sure what I'm missing.

**Comment by larsks** [comment]
@boca, you're looking at the wrong PR.  See #1744.

**Comment by bcoca** [comment]
i'm going to assume that means I can close this one
